cpp/test/sg/tsne_test.cu

```diff
@@ -115,7 +117,105 @@ class TSNETest : public ::testing::Test {
     free(embeddings_h);
   }
 
-  void SetUp() override { basicTest(); }
+  void fitWithKNNTest() {
+    cumlHandle handle;
+
+    // Allocate memory
+    device_buffer<float> X_d(handle.getDeviceAllocator(), handle.getStream(),
+                             n * p);
+    MLCommon::updateDevice(X_d.data(), digits.data(), n * p,
+                           handle.getStream());
+    CUDA_CHECK(cudaStreamSynchronize(handle.getStream()));
+
+    device_buffer<float> Y_d(handle.getDeviceAllocator(), handle.getStream(),
+                             n * 2);
+
+    MLCommon::device_buffer<int64_t> knn_indices(handle.getDeviceAllocator(),
+                                                 handle.getStream(), n * 90);
+
+    CUDA_CHECK(cudaStreamSynchronize(handle.getStream()));
+
+    MLCommon::device_buffer<float> knn_dists(handle.getDeviceAllocator(),
+                                             handle.getStream(), n * 90);
+
+    CUDA_CHECK(cudaStreamSynchronize(handle.getStream()));
+
+    TSNE::get_distances(X_d.data(), n, p, knn_indices.data(), knn_dists.data(),
+                        90, handle.getDeviceAllocator(), handle.getStream());
+
+    CUDA_CHECK(cudaStreamSynchronize(handle.getStream()));
+
+    // Test Barnes Hut
+    TSNE_fit(handle, X_d.data(), Y_d.data(), n, p, knn_indices.data(),
+             knn_dists.data(), 2, 90, 0.5, 0.0025, 50, 100, 1e-5, 12, 250, 0.01,
+             200, 500, 1000, 1e-7, 0.5, 0.8, -1);
+
+    // Move embeddings to host.
+    // This can be used for printing if needed.
+    float *embeddings_h = (float *)malloc(sizeof(float) * n * 2);
+    assert(embeddings_h != NULL);
+
+    MLCommon::updateHost(&embeddings_h[0], Y_d.data(), n * 2,
+                         handle.getStream());
+    CUDA_CHECK(cudaStreamSynchronize(handle.getStream()));
+
+    // Transpose the data
+    int k = 0;
+    float C_contiguous_embedding[n * 2];
+    for (int i = 0; i < n; i++) {
+      for (int j = 0; j < 2; j++)
+        C_contiguous_embedding[k++] = embeddings_h[j * n + i];
+    }
+
+    // Move transposed embeddings back to device, as trustworthiness requires C contiguous format
+    MLCommon::updateDevice(Y_d.data(), C_contiguous_embedding, n * 2,
+                           handle.getStream());
+    CUDA_CHECK(cudaStreamSynchronize(handle.getStream()));
```

Comment:
This is just a nitpick since it's a test, but you only need the explicit synchronize when copying to host. You don't need to do it when copying to device.